micropython/lora/lora/sx1276.py

```diff
@@ -0,0 +1,592 @@
+            update_mask |= _INVERT_IQ_TX
+        except KeyError:
+            pass
+        self._reg_update(_REG_INVERT_IQ, update_mask, invert_iq)
+
+        # TODO: Set Low Data Rate Optimization based on symbol rate (required if symbol duration > 16ms), semtech driver
+        # seems to set this for BW<=250kHz (TODO check)
+
+        try:
+            preamble_len = lora_cfg["preamble_len"]
+            preamble_len = bytes([preamble_len >> 8, preamble_len])
+            self._reg_write(_REG_PREAMBLE_LEN_MSB, preamble_len)
+        except KeyError:
+            pass
+
+        # TODO: make AGC configurable?
+        self._reg_write(_REG_MODEM_CONFIG3, _MODEM_CONFIG3_AGC_ON)
+
+        # Come out of sleep mode
+        self._set_mode(_OPMODE_MODE_STDBY)
+
+    def _reg_write(self, reg, value):
+        self._cs(0)
+        if isinstance(value, int):
+            self._buf2[0] = reg | _WRITE_REG_BIT
+            self._buf2[1] = value
+            self._spi.write(self._buf2)
+            if _DEBUG:
+                dbg = hex(value)
+        else:  # value is a buffer
+            self._buf1[0] = reg | _WRITE_REG_BIT
+            self._spi.write(self._buf1)
+            self._spi.write(value)
+            if _DEBUG:
+                dbg = binascii.hexlify(value)
+        self._cs(1)
+
+        if _DEBUG:
+            print("W {:#x} ==> {}".format(reg, dbg))
+            self._reg_read(reg)  # log the readback as well
+
+    def _reg_update(self, reg, update_mask, new_value):
+        # Note: this function has a race condition, if the register value changes between
+        # read and write then old_value will be stale
+        old_value = self._reg_read(reg)
+        if update_mask:
+            value = ((old_value & ~update_mask) & 0xFF) | (new_value & update_mask)
+            self._reg_write(reg, value)
+        return old_value
+
+    def _reg_read(self, reg):
+        self._buf2[0] = reg
+        self._buf2[1] = 0xFF
+        self._cs(0)
+        self._spi.write_readinto(self._buf2, self._buf2)
+        self._cs(1)
+        if _DEBUG:
+            print("R {:#x} <== {:#x}".format(reg, self._buf2[1]))
+        return self._buf2[1]
+
+    def _reg_readinto(self, reg, buf):
+        self._cs(0)
+        self._spi.readinto(self._buf1, reg)
+        self._spi.readinto(buf)
+        if _DEBUG:
+            print("R {:#x} <== {}".format(reg, binascii.hexlify(buf)))
+        self._cs(1)
+
+    def _get_mode(self):
+        return self._reg_read(_REG_OPMODE) & _OPMODE_MODE_MASK
+
+    def _set_mode(self, mode, lora_en=True):
+        lora_val = _OPMODE_LONGRANGEMODE_LORA if lora_en else 0
+        old_value = self._reg_update(
+            _REG_OPMODE, _OPMODE_LONGRANGEMODE_LORA | _OPMODE_MODE_MASK, mode | lora_val
+        )
+        old_value &= _OPMODE_MODE_MASK
+        if old_value == _OPMODE_MODE_SLEEP:
+            time.sleep_ms(1)  # waking from sleep takes TS_OSC = 250us (typical)
```

Comment:
why is this delay needed?

Reply:
I think this can be sleep_us(250), will change. Somehow I missed that API call was available.

(My understanding of TS_OSC is that the modem won't accept any further SPI commands from the host for minimum of this long after coming out of sleep mode. Is that wrong?)

Reply:
Where did you see that in the datasheet? I don't think I have used sleep mode on the sx1276, so I don't have any off-datasheet experience.

Reply:
You're right, I don't know why I thought this was necessary. Looks like it's only necessary for FIFO access over SPI, and I don't think this will happen within TS_OSC after coming out of sleep. Will take it out.